smolyay/basis.py

```diff
@@ -0,0 +1,226 @@
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        if self._n == 0:
+            return 1
+        elif self._n == 1:
+            return x
+        else:
+            k_lim = self._n//2
+            answer = 0
+            for k in range(0,k_lim+1):
+                answer += math.comb(self._n,2*k)*((x**2 - 1)**k)*(x**(self._n-2*k))
+            return answer
+
+
+class BasisFunctionSet():
+    """Set of basis functions
+    Creates a framework for a set of basis functions to be used as the
+    building blocks for a surrogate model.
+    ``basis_set`` is a list of BasisFunction objects
+    ''all_points'' 1D points taken from the BasisFunction objects
+
+    """
+
+    def __init__(self,all_points,basis_set):
+        self._basis_set = basis_set
+        self._all_points = all_points
+
+    @property
+    def all_points(self):
+        """All points for basis function set at some level of precision"""
+        return self._all_points
+
+    @all_points.setter
+    def all_points(self,all_points):
+        self._all_points = all_points
+
+    @property
+    def basis_set(self):
+        """list of BasisFunction objects"""
+        return self._basis_set
+
+    @basis_set.setter
+    def basis_set(self,basis_set):
+        self._basis_set = basis_set
+
+
+class NestedBasisFunctionSet(BasisFunctionSet):
+    """Set of basis functions that are nested
+
+    The set of basis functions this class uses are assumed to be part
+    of a sequential series of functions, where the functions that
+    determine the points are nested
+    ``levels`` is a list of lists that determines which grid level each
+    point is added on
+    """
+
+    def __init__(self,all_points,basis_set,levels):
+        """Initialization of parameters"""
+        super().__init__(all_points,basis_set)
+        self._levels = levels
+
+    @property
+    def levels(self):
+        """list of list that show the points for each grid level"""
+        return self._levels
+
+    @levels.setter
+    def levels(self,levels):
+        self._levels = levels
+
+
+def make_nested_chebyshev_points(exactness,basis_function):
+    """calculate nested points for Chebyshev polynomial basis function
+    Created the NestedBasisFunctionSet object that holds the nested
+    points for a CHebyshev basis function
+
+    Parameters
+    ----------
+    exactness : int
+        level of exactness to calculate points to
+    basis_function : BasisFunction object
+        type of chebyshev polynomial
+
+    Returns
+    -------
+    nested_set : NestedBasisFunctionSet object
+        data structure for points
+    """
+
+    levels = [[0]]
+    basis_set = []
+    all_points = [0]
+    max_degree = 2**exactness
+    if exactness == 0:
+        max_degree = 0
+    start_level_index = 1
+    for i in range(0,max_degree+1):
+        basis_set.append(basis_function(i))
```

Comment:
This can be done in one line!
```suggestion
    basis_set = [basis_function(n) for n in range(max_degree+1)]
```